lib/Sema/MiscDiagnostics.cpp

```diff
@@ -3114,7 +3114,26 @@ class ObjCSelectorWalker : public ASTWalker {
           name = bestMethod->getFullName();
         }
 
-        out << nominal->getName().str() << "." << name.getBaseName().str();
+        // Only print the type name if the function is implemented outside the
+        // current class.
+        {
+          auto shouldAddNominalTypeName = [&]() {
```

Comment:
Please make this a new static function -- the method this logic is nested inside is already too long